File: bindings/csharp/Hedl.Tests/Fixtures.cs
// Common test fixtures loader for HEDL C# bindings.
//
// This class provides access to shared test fixtures stored in the
// bindings/common/fixtures directory, eliminating test data duplication
// across language bindings.

using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;

namespace Dweve.Hedl.Tests
{
    /// <summary>
    /// Manifest structure for fixture definitions.
    /// </summary>
    internal class FixtureManifest
    {
        public Dictionary<string, FixtureEntry> Fixtures { get; set; } = new();
        public Dictionary<string, ErrorEntry> Errors { get; set; } = new();
    }

    /// <summary>
    /// Fixture entry in the manifest.
    /// </summary>
    internal class FixtureEntry
    {
        public string Description { get; set; } = string.Empty;
        public Dictionary<string, string> Files { get; set; } = new();
    }

    /// <summary>
    /// Error fixture entry in the manifest.
    /// </summary>
    internal class ErrorEntry
    {
        public string Description { get; set; } = string.Empty;
        public string File { get; set; } = string.Empty;
        public bool ExpectedError { get; set; }
    }

    /// <summary>
    /// Loads and provides access to common HEDL test fixtures.
    ///
    /// All fixtures are loaded from bindings/common/fixtures directory
    /// to ensure consistency across language bindings.
    /// </summary>
    public class Fixtures
    {
        private readonly string _fixturesDir;
        private readonly FixtureManifest _manifest;

        /// <summary>
        /// Initialize the fixtures loader and load manifest.
        /// </summary>
        public Fixtures()
        {
            // Path to common fixtures directory
            // Navigate from bin/Debug/net6.0 (or similar) to common/fixtures
            var baseDir = AppDomain.CurrentDomain.BaseDirectory;
            _fixturesDir = Path.GetFullPath(
                Path.Combine(baseDir, "..", "..", "..", "..", "..", "common", "fixtures")
            );

            // Load manifest
            var manifestPath = Path.Combine(_fixturesDir, "manifest.json");
            if (!File.Exists(manifestPath))
            {
                throw new FileNotFoundException($"Manifest not found at: {manifestPath}");
            }

            var manifestContent = File.ReadAllText(manifestPath);
            var options = new JsonSerializerOptions
            {
                PropertyNamingPolicy = JsonNamingPolicy.CamelCase
            };
            _manifest = JsonSerializer.Deserialize<FixtureManifest>(manifestContent, options)
                ?? throw new InvalidOperationException("Failed to parse manifest");
        }

        /// <summary>
        /// Read a fixture file.
        /// </summary>
        /// <param name="filename">Name of the file to read</param>
        /// <returns>File contents as string</returns>
        private string ReadFile(string filename)
        {
            var filepath = Path.Combine(_fixturesDir, filename);
            if (!File.Exists(filepath))
            {
                throw new FileNotFoundException($"Fixture file not found: {filepath}");
            }

            return File.ReadAllText(filepath);
        }

        // Basic fixtures

        /// <summary>
        /// Get basic HEDL sample document.
        /// </summary>
        public string BasicHedl => ReadFile(_manifest.Fixtures["basic"].Files["hedl"]);

        /// <summary>
        /// Get basic JSON sample document.
        /// </summary>
        public string BasicJson => ReadFile(_manifest.Fixtures["basic"].Files["json"]);

        /// <summary>
        /// Get basic YAML sample document.
        /// </summary>
        public string BasicYaml => ReadFile(_manifest.Fixtures["basic"].Files["yaml"]);

        /// <summary>
        /// Get basic XML sample document.
        /// </summary>
        public string BasicXml => ReadFile(_manifest.Fixtures["basic"].Files["xml"]);

        // Type-specific fixtures

        /// <summary>
        /// Get HEDL document with various scalar types.
        /// </summary>
        public string ScalarsHedl => ReadFile(_manifest.Fixtures["scalars"].Files["hedl"]);

        /// <summary>
        /// Get HEDL document with nested structures.
        /// </summary>
        public string NestedHedl => ReadFile(_manifest.Fixtures["nested"].Files["hedl"]);

        /// <summary>
        /// Get HEDL document with lists and arrays.
        /// </summary>
        public string ListsHedl => ReadFile(_manifest.Fixtures["lists"].Files["hedl"]);

        // Performance fixtures

        /// <summary>
        /// Get large HEDL document for performance testing.
        /// </summary>
        public string LargeHedl => ReadFile(_manifest.Fixtures["large"].Files["hedl"]);

        // Error fixtures

        /// <summary>
        /// Get invalid HEDL syntax for error testing.
        /// </summary>
        public string ErrorInvalidSyntax => ReadFile(_manifest.Errors["invalid_syntax"].File);

        /// <summary>
        /// Get malformed HEDL document for error testing.
        /// </summary>
        public string ErrorMalformed => ReadFile(_manifest.Errors["malformed"].File);

        // Utility methods

        /// <summary>
        /// Get a specific fixture by category and format.
        /// </summary>
        /// <param name="category">Fixture category ("basic", "scalars", etc.)</param>
        /// <param name="format">File format ("hedl", "json", "yaml", "xml")</param>
        /// <returns>Fixture content as string</returns>
        /// <example>
        /// <code>
        /// var fixtures = new Fixtures();
        /// var hedl = fixtures.GetFixture("basic", "hedl");
        /// </code>
        /// </example>
        public string GetFixture(string category, string format = "hedl")
        {
            if (_manifest.Fixtures.TryGetValue(category, out var entry))
            {
                if (entry.Files.TryGetValue(format, out var filename))
                {
                    return ReadFile(filename);
                }
            }

            throw new ArgumentException($"Fixture not found: category={category}, format={format}");
        }

        /// <summary>
        /// Get an error fixture by type.
        /// </summary>
        /// <param name="errorType">Type of error ("invalid_syntax", "malformed")</param>
        /// <returns>Error fixture content</returns>
        public string GetErrorFixture(string errorType)
        {
            if (_manifest.Errors.TryGetValue(errorType, out var entry))
            {
                return ReadFile(entry.File);
            }

            throw new ArgumentException($"Error fixture not found: {errorType}");
        }
    }

    /// <summary>
    /// Global fixtures instance for convenient access.
    /// </summary>
    public static class GlobalFixtures
    {
        private static readonly Lazy<Fixtures> _instance = new(() => new Fixtures());

        /// <summary>
        /// Get the global fixtures instance.
        /// </summary>
        public static Fixtures Instance => _instance.Value;

        // Legacy constants for backward compatibility
        public static string SampleHedl => Instance.BasicHedl;
        public static string SampleJson => Instance.BasicJson;
        public static string SampleYaml => Instance.BasicYaml;
        public static string SampleXml => Instance.BasicXml;
    }
}

File: bindings/go/fixtures.go
// Package hedl provides test fixtures for HEDL Go bindings.
//
// This package provides access to shared test fixtures stored in the
// bindings/common/fixtures directory, eliminating test data duplication
// across language bindings.
package hedl

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
)

// FixtureManifest represents the structure of the manifest.json file.
type FixtureManifest struct {
	Fixtures map[string]FixtureEntry `json:"fixtures"`
	Errors   map[string]ErrorEntry   `json:"errors"`
}

// FixtureEntry represents a fixture entry in the manifest.
type FixtureEntry struct {
	Description string            `json:"description"`
	Files       map[string]string `json:"files"`
}

// ErrorEntry represents an error fixture entry in the manifest.
type ErrorEntry struct {
	Description   string `json:"description"`
	File          string `json:"file"`
	ExpectedError bool   `json:"expected_error"`
}

// Fixtures provides access to common HEDL test fixtures.
//
// All fixtures are loaded from bindings/common/fixtures directory
// to ensure consistency across language bindings.
type Fixtures struct {
	fixturesDir string
	manifest    FixtureManifest
}

// NewFixtures creates a new Fixtures instance and loads the manifest.
func NewFixtures() (*Fixtures, error) {
	// Get the directory of this source file
	_, filename, _, ok := runtime.Caller(0)
	if !ok {
		return nil, &HedlError{Message: "failed to get caller information"}
	}

	// Path to common fixtures directory
	// From bindings/go/fixtures.go -> bindings/common/fixtures
	fixturesDir := filepath.Join(filepath.Dir(filename), "..", "common", "fixtures")

	// Load manifest
	manifestPath := filepath.Join(fixturesDir, "manifest.json")
	manifestData, err := os.ReadFile(manifestPath)
	if err != nil {
		return nil, &HedlError{Message: "failed to read manifest: " + err.Error()}
	}

	var manifest FixtureManifest
	if err := json.Unmarshal(manifestData, &manifest); err != nil {
		return nil, &HedlError{Message: "failed to parse manifest: " + err.Error()}
	}

	return &Fixtures{
		fixturesDir: fixturesDir,
		manifest:    manifest,
	}, nil
}

// readFile reads a fixture file and returns its contents.
func (f *Fixtures) readFile(filename string) (string, error) {
	filepath := filepath.Join(f.fixturesDir, filename)
	data, err := os.ReadFile(filepath)
	if err != nil {
		return "", &HedlError{Message: "failed to read fixture: " + err.Error()}
	}
	return string(data), nil
}

// Basic fixtures

// BasicHEDL returns the basic HEDL sample document.
func (f *Fixtures) BasicHEDL() (string, error) {
	return f.readFile(f.manifest.Fixtures["basic"].Files["hedl"])
}

// BasicJSON returns the basic JSON sample document.
func (f *Fixtures) BasicJSON() (string, error) {
	return f.readFile(f.manifest.Fixtures["basic"].Files["json"])
}

// BasicYAML returns the basic YAML sample document.
func (f *Fixtures) BasicYAML() (string, error) {
	return f.readFile(f.manifest.Fixtures["basic"].Files["yaml"])
}

// BasicXML returns the basic XML sample document.
func (f *Fixtures) BasicXML() (string, error) {
	return f.readFile(f.manifest.Fixtures["basic"].Files["xml"])
}

// Type-specific fixtures

// ScalarsHEDL returns a HEDL document with various scalar types.
func (f *Fixtures) ScalarsHEDL() (string, error) {
	return f.readFile(f.manifest.Fixtures["scalars"].Files["hedl"])
}

// NestedHEDL returns a HEDL document with nested structures.
func (f *Fixtures) NestedHEDL() (string, error) {
	return f.readFile(f.manifest.Fixtures["nested"].Files["hedl"])
}

// ListsHEDL returns a HEDL document with lists and arrays.
func (f *Fixtures) ListsHEDL() (string, error) {
	return f.readFile(f.manifest.Fixtures["lists"].Files["hedl"])
}

// Performance fixtures

// LargeHEDL returns a large HEDL document for performance testing.
func (f *Fixtures) LargeHEDL() (string, error) {
	return f.readFile(f.manifest.Fixtures["large"].Files["hedl"])
}

// Error fixtures

// ErrorInvalidSyntax returns invalid HEDL syntax for error testing.
func (f *Fixtures) ErrorInvalidSyntax() (string, error) {
	return f.readFile(f.manifest.Errors["invalid_syntax"].File)
}

// ErrorMalformed returns a malformed HEDL document for error testing.
func (f *Fixtures) ErrorMalformed() (string, error) {
	return f.readFile(f.manifest.Errors["malformed"].File)
}

// Utility methods

// GetFixture returns a specific fixture by category and format.
//
// Example:
//
//	fixtures, _ := NewFixtures()
//	hedl, _ := fixtures.GetFixture("basic", "hedl")
func (f *Fixtures) GetFixture(category, format string) (string, error) {
	if entry, ok := f.manifest.Fixtures[category]; ok {
		if filename, ok := entry.Files[format]; ok {
			return f.readFile(filename)
		}
	}

	return "", &HedlError{Message: "fixture not found: category=" + category + ", format=" + format}
}

// GetErrorFixture returns an error fixture by type.
func (f *Fixtures) GetErrorFixture(errorType string) (string, error) {
	if entry, ok := f.manifest.Errors[errorType]; ok {
		return f.readFile(entry.File)
	}

	return "", &HedlError{Message: "error fixture not found: " + errorType}
}

// Global fixtures instance for convenient access
var globalFixtures *Fixtures

func init() {
	var err error
	globalFixtures, err = NewFixtures()
	if err != nil {
		panic("failed to initialize fixtures: " + err.Error())
	}
}

// GetGlobalFixtures returns the global fixtures instance.
func GetGlobalFixtures() *Fixtures {
	return globalFixtures
}
